Revisions for Manuscript (#24)
greenelab/text_mined_hetnet_manuscript: content/03.methods.md

### Hetionet

![

> A metagraph (schema) of Hetionet where pharmacological, biological and disease entities are represented as nodes and the relationships between them are represented as edges.

Reviewer: ```suggestion
A metagraph (schema) of Hetionet where biomedical entities are represented as nodes and the relationships between them are represented as edges.
```

> This project only focuses on the information shown in bold; however, we can extend this work to incorporate the faded out information as well.

Reviewer: ```suggestion
We examined performance on the highlighted subgraph; however, the long-term vision is to capture edges for the entire graph.
```

PubTator provides MEDLINE abstracts that have been annotated with well-established entity recognition tools including DNorm [@doi:10.1093/bioinformatics/btt474] for disease mentions, GeneTUKit [@doi:10.1093/bioinformatics/btr042] for gene mentions, Gnorm [@doi:10.1186/1471-2105-12-S8-S5] for gene normalizations and a dictionary based search system for compound mentions [@doi:10.1093/database/bas037].
We downloaded PubTator on June 30, 2017, at which point it contained 10,775,748 abstracts. 
Then we filtered out mention tags that were not contained in hetionet.
We used the Stanford CoreNLP parser [@doi:10.3115/v1/P14-5010] to tag parts of speech and generate dependency trees.
We extracted sentences with two or more mentions, termed candidate sentences.
Each candidate sentence was stratified by co-mention pair to produce a training set, tuning set and a testing set (shown in Table {@tbl:candidate-sentences}).
Each unique co-mention pair is sorted into four categories: (1) in hetionet and has sentences, (2) in hetionet and doesn't have sentences, (3) not in hetionet and does have sentences and (4) not in hetionet and doesn't have sentences.

> Within these four categories each pair receives their own individual partition rank (continuous number between 0 and 1).

Reviewer: Randomly assigned?

Author: Yeah the partition rank is randomly assigned per group by group basis.

Reviewer: This would be a good thing to specify.

Any rank lower than 0.7 is sorted into the training set, while any rank greater than 0.7 and lower than 0.9 is assigned to the tuning set.
The rest of the pairs with a rank greater than or equal to 0.9 is assigned to the test set.
Sentences that contain more than one co-mention pair are treated as multiple individual candidates.

> We hand labeled five hundred to a thousand candidate sentences of each relationship to obtain a ground truth set (Table {@tbl:candidate-sentences})[^1].

Reviewer: ```suggestion
We hand labeled five hundred to a thousand candidate sentences of each relationship type to obtain a ground truth set (Table {@tbl:candidate-sentences})[^1].
```

### Label Functions for Annotating Sentences

> A common challenge in natural language processing is having a miniscule amount of ground truth annotations, even when textual data are abundant.

Reviewer: ```suggestion
The challenge of having too few ground truth annotations is common to many natural language processing settings, even when unannotated text is abundant.
```

**Databases**: These label functions incorporate existing databases to generate a signal, as seen in distant supervision [@doi:10.3115/1690219.1690287].
These functions detect if a candidate sentence's co-mention pair is present in a given database.

> If the candidate pair is present, our label function will emit a positive label and abstain otherwise.

Reviewer: ```suggestion
If the candidate pair is present, our label function emitted a positive label and abstained otherwise.
```

> If the candidate pair isn't present in any existing database, then a separate label function will emit a negative label.

Reviewer: ```suggestion
If the candidate pair wasn't present in any existing database, a separate label function emitted a negative label.
```

> We used a separate label function to prevent a label imbalance problem, which occurred when candidates that rarely appeared in databases were drowned out by negative labels.

Reviewer: ```suggestion
We used a separate label function to prevent a label imbalance problem that we encountered during development: emitting positive and negatives from the same label functions appeared to result in classifiers that predict almost exclusively negative predictions.
```

> We use a separate label function to prevent a label imbalance problem. This problem occurs when candidates, that scarcely appear in databases, are drowned out by negative labels.

Reviewer: ```suggestion
We used a separate label function to prevent a label imbalance problem, which occurred when candidates that rarely appeared in databases were drowned out by negative labels.
```
I'm still not sure what you mean here, but the change fixes one sentence per line.

Author: I was trying to provide rationale for using a separate label function to emit a negative label instead of having all distant supervision functions emit both positive and negative labels. The latter case breaks the generative model and ends up making the classifier predict negative more than desired.

Reviewer: Ok! I tried to write this rationale in to a revised suggestion. See what you think.

> The multitude of negative labels increases the likelihood of misclassification when training the generative model.

Reviewer: ```suggestion
```

$$\hat{\theta} = argmin_{\theta} -\sum_{\Lambda} log \sum_{Y}P(\Lambda, Y)$$

In the framework we used predictions from the generative model, $\hat{Y} = P(Y \mid \Lambda)$, as training classes for our dataset [@doi:10.14778/3157794.3157797; @doi:10.1145/3209889.3209898].

#### Word Embeddings

Word embeddings are representations that map individual words to real valued vectors of user-specified dimensions.
These embeddings have been shown to capture the semantic and syntactic information between words [@arxiv:1310.4546].

> Using all candidate sentences for each individual relationship pair, we trained Facebook's fastText [@arxiv:1607.04606] to generate word embeddings.

Reviewer: ```suggestion
We trained Facebook's fastText [@arxiv:1607.04606] using all candidate sentences for each individual relationship pair to generate word embeddings.
```

> The fastText model uses a skipgram model [@arxiv:1301.3781] that aims to predict the context given a candidate word and pairs the model with a novel scoring function that treats each word as a bag of character n-grams.

Reviewer: ```suggestion
fastText uses a skipgram model [@arxiv:1301.3781] that aims to predict the surrounding context for a candidate word and pairs the model with a novel scoring function that treats each word as a bag of character n-grams.
```

We trained this model for 20 epochs using a window size of 2 and generated 300-dimensional word embeddings.
We use the optimized word embeddings to train a discriminative model.  

#### Discriminative Model

The discriminative model is a neural network, which we train to predict labels from the generative model.
The expectation is that the discriminative model can learn more complete features of the text than the label functions used in the generative model.
We used a convolutional neural network with multiple filters as our discriminative model.
This network uses multiple filters with fixed widths of 300 dimensions and a fixed height of 7 (Figure {@fig:convolutional_network}), because this height provided the best performance in terms of relationship classification [@arxiv:1510.03820].
We trained this model for 20 epochs using the adam optimizer [@arxiv:1412.6980] with pytorch's default parameter settings and a learning rate of 0.001.
We added a L2 penalty on the network weights to prevent overfitting.
Lastly, we added a dropout layer (p=0.25) between the fully connected layer and the softmax layer.

![

> The architecture of the discriminative model is a convolutional neural network.

Reviewer: ```suggestion
The architecture of the discriminative model was a convolutional neural network.
```

> We perform a convolution step using multiple filters. 

Reviewer: ```suggestion
We performed a convolution step using multiple filters. 
```

> These filters generate a feature map that is sent into a maximum pooling layer that is designed to extract the largest feature in each map.

Reviewer: ```suggestion
The filters generated a feature map that was sent into a maximum pooling layer that was designed to extract the largest feature in each map.
```

> The extracted features are concatenated into a singular vector that is passed into a fully connected network. 

Reviewer: ```suggestion
The extracted features were concatenated into a singular vector that was passed into a fully connected network. 
```

> The fully connected network has 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 

Reviewer: ```suggestion
The fully connected network had 300 neurons for the first layer, 100 neurons for the second layer and 50 neurons for the last layer. 
```

> From the fully connected network the last step is to generate predictions using a softmax layer.

Reviewer: ```suggestion
The last step from the fully connected network was to generate predictions using a softmax layer.
```

Using a sampling with replacement approach, we sampled these text and domain-heuristic label functions separately within edge types, across edge types, and from a pool of all label functions.
We compared within-edge-type performance to across-edge-type and all-edge-type performance.
For each edge type we sampled a fixed number of label functions consisting of five evenly-spaced numbers between one and the total number of possible label functions.
We repeated this sampling process 50 times for each point.
We evaluated both generative and discriminative models at each point, and we reported performance of each in terms of the area under the receiver operating characteristic curve (AUROC) and the area under the precision-recall curve (AUPR).

#### Adding Random Noise to Generative Model

We discovered in the course of this work that adding a single label function from a mismatched type would often improve the performance of the generative model (see Results).
We designed an experiment to test whether adding a noisy label function also increased performance.
This label function emitted a positive or negative label at varying frequencies, which were evenly spaced from zero to one.

> Zero being the same as distant supervision and one meaning all sentences are labeled.

Reviewer: ```suggestion
Zero was the same as distant supervision and one meant that all sentences were randomly labeled.
```
